envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt

```diff
@@ -23,27 +29,36 @@ class SnapshotDebugController(controlPlane: ControlPlane) {
      * extracted from Envoy's config_dump endpoint.
      */
     @PostMapping("/snapshot")
-    fun snapshot(@RequestBody node: Node): ResponseEntity<String> {
+    fun snapshot(@RequestBody node: Node): ResponseEntity<SnapshotDebugInfo> {
         val nodeHash = nodeGroup.hash(node)
         val snapshot = cache.getSnapshot(nodeHash)
         return if (snapshot == null) {
-            return ResponseEntity("snapshot missing", HttpStatus.NOT_FOUND)
+            throw ResponseStatusException(HttpStatus.NOT_FOUND, "snapshot missing")
         } else {
             ResponseEntity(
-                versions(snapshot) +
-                    "snapshot:\n" +
-                    snapshot.toString(),
+                SnapshotDebugInfo(
+                    snapshot = Snapshot(
+                        clusters = snapshot.clusters().resources(),
+                        endpoints = snapshot.endpoints().resources(),
+                        listeners = snapshot.listeners().resources(),
+                        routes = snapshot.routes().resources()
+                    ),
+                    versions = Versions(
+                        clusters = version(snapshot.clusters().version()),
+                        endpoints = version(snapshot.endpoints().version()),
+                        listeners = version(snapshot.listeners().version()),
+                        routes = version(snapshot.routes().version())
+                    )
+                ),
                 HttpStatus.OK
             )
         }
     }
 
-    private fun versions(snapshot: Snapshot): String {
-        val versions = StringBuilder()
-            .append("clusters: ${snapshot.clusters().version()}\n")
-            .append("endpoints: ${snapshot.endpoints().version()}\n")
-            .append("routes: ${snapshot.routes().version()}\n")
-            .append("listeners: ${snapshot.listeners().version()}\n")
-        return versions.toString()
+    @JsonComponent
+    class ProtoSerializer : JsonSerializer<Message>() {
```

Comment:
Shouldn't  `protobufJsonFormatHttpMessageConverter` from `ControlPlaneConfig` take care of it?

Reply:
unfortunatelly `protobufJsonFormatHttpMessageConverter` handles only deserialisation, not serialisation.